### PR TITLE
Use `str{l,n}cpy()` in `rz_str_ncpy()` instead of a loop

### PR DESCRIPTION
**Pull request by officialcjunior**
# SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

See https://github.com/rizinorg/rizin-extras/pull/4#discussion_r680669470.

The current implementation of `rz_strn_cpy()` uses a `for` loop and iterates through the whole array `n` times
and copies it. This can be made more efficient by using `memcpy` in it, instead of the loop.

**Test plan**

CI is green

**Closing issues**

None.

### CODE REVIEW

File: librz/util/str.c [hunk 1]
```diff
@@ -764,10 +764,8 @@ RZ_API size_t rz_str_ncpy(char *dst, const char *src, size_t n) {
 	}
 
 	n--;
-	for (i = 0; src[i] && n > 0; i++, n--) {
-		dst[i] = src[i];
-	}
-	dst[i] = 0;
+	memcpy(dst, src, n);
```
Comment (XVilka) [inline librz/util/str.c]: If n == 1, it will call `memcpy(dst, src, 0)` which is wrong.

Comment (XVilka) [reply librz/util/str.c]: It should be 
```c
memcpy(dst, src, n);
dst[n-1] = '\0';
```
and remove `n--` before.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
Not sure why it broke these particular test on Windows but not on the other systems:
```
[XX] C:\projects\rizin\test\db\formats\mach0\mach0 osx mach0 entrypoint
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc ie~addr bins/mach0/ls-osx-x86_64
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-vaddr=0x1000013d8 paddr=0x000013d8 haddr=0x00000648 type=program
-- exit status: -1073741819

[XX] C:\projects\rizin\test\db\cmd\feat_foreach @@c:/c/a fs:
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e search.in=bin.sections
?v $$ @@c:/ao fs:[ra
' bins/mach0/mac-ls
-- stdout
--- expected
+++ actual
@@ -1,4 +1,0 @@
-0x100004cd7
-0x100004d95
-0x100004e1d
-0x100005078
-- exit status: -1073741819

[XX] C:\projects\rizin\test\db\cmd\cmd_search /aI infinite loop instructions
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc /aI bins/mach0/ls-osx-x86_64
-- stdout
--- expected
+++ actual
@@ -1,5 +1,0 @@
-0x10000166f
-0x10000221d
-0x1000047ca
-0x10000482e
-0x100004838
-- exit status: -1073741819

[XX] C:\projects\rizin\test\db\cmd\cmd_i iz (__cfstring)
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?e -- default --
iz
?e -- geq 12 --
e bin.minstr=12
iz
?e -- leq 13 --
e bin.maxstr=12
iz
' bins/mach0/hello-objc-osx
-- stdout
--- expected
+++ actual
@@ -1,34 +1,0 @@
--- default --
-[Strings]
-nth paddr       vaddr       len size section            type  string            
---------------------------------------------------------------------------------
-0   0x00000d58  0x100000d58 9   10   2.__TEXT.__cstring ascii sayHello:
-1   0x00000d62  0x100000d62 4   5    2.__TEXT.__cstring ascii init
-2   0x00000d67  0x100000d67 5   6    2.__TEXT.__cstring ascii alloc
-3   0x00000d6d  0x100000d6d 5   6    2.__TEXT.__cstring ascii Hahha
-4   0x00000d73  0x100000d73 12  13   2.__TEXT.__cstring ascii Hello World 
-5   0x00000d86  0x100000d86 9   10   2.__TEXT.__cstring ascii callMeNot
-6   0x00000d90  0x100000d90 7   8    2.__TEXT.__cstring ascii v16@0:8
-7   0x00000d98  0x100000d98 9   10   2.__TEXT.__cstring ascii sayHello2
-8   0x00000da2  0x100000da2 10  11   2.__TEXT.__cstring ascii v20@0:8i16
-9   0x00000dad  0x100000dad 11  12   2.__TEXT.__cstring ascii sayHello3::
-10  0x00000db9  0x100000db9 13  14   2.__TEXT.__cstring ascii v24@0:8i16i20
-11  0x00000dc7  0x100000dc7 8   9    2.__TEXT.__cstring ascii sayLong:
-12  0x00000dd0  0x100000dd0 10  11   2.__TEXT.__cstring ascii v24@0:8Q16
-13  0x00000ddb  0x100000ddb 7   8    2.__TEXT.__cstring ascii @16@0:8
-14  0x00000de3  0x100000de3 4   5    2.__TEXT.__cstring ascii Test
-3   0x100001058 0x100001058 5   6                       ascii cstr.Hahha
-4   0x100001078 0x100001078 12  13                      ascii cstr.Hello World 
--- geq 12 --
-[Strings]
-nth paddr       vaddr       len size section            type  string            
---------------------------------------------------------------------------------
-0   0x00000d73  0x100000d73 12  13   2.__TEXT.__cstring ascii Hello World 
-1   0x00000db9  0x100000db9 13  14   2.__TEXT.__cstring ascii v24@0:8i16i20
-0   0x100001078 0x100001078 12  13                      ascii cstr.Hello World 
--- leq 13 --
-[Strings]
-nth paddr       vaddr       len size section            type  string            
---------------------------------------------------------------------------------
-0   0x00000d73  0x100000d73 12  13   2.__TEXT.__cstring ascii Hello World 
-0   0x100001078 0x100001078 12  13                      ascii cstr.Hello World 
-- exit status: -1073741819

[XX] C:\projects\rizin\test\db\analysis\x86_64 analysis vars retype
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'af
afvt var_30h char
afvb~_30h
' bins/mach0/ls-osx-x86_64
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-var char var_30h @ rbp-0x30
-- exit status: -1073741819

[XX] C:\projects\rizin\test\db\abi\platforms\entries MAIN_ABI : x86_64-macho  entries
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 's main
s
' bins/abi_bins/elf/platforms/x86_64-macho-echo
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-0x100000e50
-- exit status: -1073741819
```

File: librz/util/str.c [hunk 2]
```diff
@@ -756,19 +756,11 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
 // Secure string copy with null terminator (like strlcpy or strscpy but ours
 RZ_API size_t rz_str_ncpy(char *dst, const char *src, size_t n) {
 	rz_return_val_if_fail(dst && src, 0);
-	size_t i;
 
-	// do not do anything if n is 0
```
Comment (XVilka) [inline librz/util/str.c:761]: I think it's better to still keep the `0` check.

Comment (officialcjunior) [reply librz/util/str.c:761]: I'll try bringing back the same check, then.

File: librz/util/str.c [hunk 3]
```diff
@@ -756,19 +756,11 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
...
-	if (n == 0) {
-		return 0;
-	}
+	memcpy(dst, src, n);
```
Comment (ret2libc) [inline librz/util/str.c]: This is not exactly the same as before, because before there was the check `src[i]`, which disappear if you use `memcpy`.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/util/str.c (hunk 3 above)
Comment (wargio) [reply librz/util/str.c]: i guess we could use `strncpy` instead of `memcpy` so it gets optimized by the compiler

Comment (XVilka) [reply librz/util/str.c]: Good idea, we could remove everything and just keep `strncpy()` here.

**Comment by XVilka** [comment]
Curiously, seems `strncpy()` is already widely used in Rizin code. I wonder, maybe it makes sense to remove this function completely? Or just substitute these `strncpy()` calls with this one? It's similar situation as with `strdup()`: https://github.com/rizinorg/rizin/issues/1176

File: librz/util/str.c (hunk 3 above)
Comment (ret2libc) [reply librz/util/str.c]: strncpy != strlcpy

**Comment by officialcjunior** [comment]
I think we can remove it, since it's basically the same thing?
I'll be happy to do it on the same PR, too.

**Comment by XVilka** [comment]
@officialcjunior it could be done this way:
```c
#if HAVE_STRLCPY
    strlcpy(dst, src, n);
#else
    strncpy(dst, src, n-1);
    dst[n] = '\0'
#endif 
```
Something like that (note I might have off-by-one error here but you got the idea).

**Comment by officialcjunior** [comment]
How do I check whether I have an off-by-one error or not?
Any particular tests I could run?

**Comment by wargio** [comment]
you can't

**Comment by ret2libc** [comment]
Well, you can create some unit tests for the function and try to test some particular cases/corner cases and compile with asan before running the tests

**Comment by XVilka** [comment]
```
test_cons_to_html ERR
[XX] Fail at line 134: Color and background: expected <font color='#f00'>he</font><font color='#f00' style='background-color:#00f'>llo</font>abc, got <font color='#f00'>he</font><font style='background-color:#00f'>llo</font>abc.
```

File: librz/util/str.c [hunk 4]
```diff
@@ -753,22 +753,22 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
...
+// Secure string copy with null terminator (like strlcpy or strscpy but ours)
 RZ_API size_t rz_str_ncpy(char *dst, const char *src, size_t n) {
 	rz_return_val_if_fail(dst && src, 0);
-	size_t i;
 
 	// do not do anything if n is 0
 	if (n == 0) {
 		return 0;
 	}
+	#if HAVE_STRLCPY
```
Comment (XVilka) [inline librz/util/str.c]: You need to add the check in the Meson buildsystem for it.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/util/str.c (hunk 4 above)
Comment (officialcjunior) [reply librz/util/str.c]: How do I do that?

Comment (XVilka) [reply librz/util/str.c]: See https://github.com/rizinorg/rizin/blob/dev/meson.build#L357

File: librz/util/str.c [hunk 5]
```diff
@@ -753,22 +753,22 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
...
 RZ_API size_t rz_str_ncpy(char *dst, const char *src, size_t n) {
 	rz_return_val_if_fail(dst && src, 0);
-	size_t i;
 
 	// do not do anything if n is 0
 	if (n == 0) {
 		return 0;
 	}
+#if HAVE_STRLCPY
+	strlcpy(dst, src, n);
```
Comment (ret2libc) [inline librz/util/str.c]: if `strlcpy` exists the whole function should just be `return strlcpy(dst, src, n);`

File: librz/util/str.c [hunk 6]
```diff
@@ -753,22 +753,22 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
 	return p;
 }
 
-// Secure string copy with null terminator (like strlcpy or strscpy but ours
+// Secure string copy with null terminator (like strlcpy or strscpy but ours)
```
Comment (ret2libc) [inline librz/util/str.c]: Use doxygen style doc.

File: test/unit/test_str.c [hunk 7]
```diff
@@ -42,6 +42,14 @@ bool test_rz_str_replace(void) {
 	mu_end;
 }
 
+bool test_rz_str_ncpy(void) {
+	char *str = strdup("hello world");
+	char buf[10];
+	rz_str_ncpy(buf, str, 3);
```
Comment (ret2libc) [inline test/unit/test_str.c:48]: please add a test for empty strings, for string of length 1, for size = 0, size=1, size=sizeof(buf), size=sizeof(buf)-1 because those may be tricky. Also, make sure to run the unit tests with an ASAN build.

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: test/unit/test_str.c (hunk 7 above)
Comment (XVilka) [reply test/unit/test_str.c:48]: See also https://github.com/rizinorg/rizin/blob/dev/test/unit/legacy_unit/util/test_str.c#L60

File: librz/util/str.c [hunk 8]
```diff
@@ -762,7 +766,7 @@ RZ_API size_t rz_str_ncpy(char *dst, const char *src, size_t n) {
 		return 0;
 	}
 #if HAVE_STRLCPY
```
Comment (wargio) [inline librz/util/str.c:768]: you can move this if after rz_str_ncpy
```c
RZ_API size_t rz_str_ncpy(char *dst, const char *src, size_t n) {
	rz_return_val_if_fail(dst && src, 0);
#if HAVE_STRLCPY
	return strlcpy(dst, src, n);
#else
	....
	return n;
#endif
}
```

**Review by wargio — CHANGES_REQUESTED** [review]

**Review by ret2libc — CHANGES_REQUESTED** [review]
Did you run the tests with ASAN as suggested? Because I see this:
```
==958462==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fffb3b5f99a at pc 0x7fca741788a0 bp 0x7fffb3b5f8b0 sp 0x7fffb3b5f060                                                   
WRITE of size 100 at 0x7fffb3b5f99a thread T0                                                                                                                                                 
    #0 0x7fca7417889f in __interceptor_strncpy (/lib64/libasan.so.6+0x5889f)                                                                                                                  
    #1 0x7fca73f65981 in rz_str_ncpy ../librz/util/str.c:771                                                                                                                                  
    #2 0x403e17 in test_rz_str_ncpy ../test/unit/test_str.c:60                                                                                                                                
    #3 0x4179db in all_tests ../test/unit/test_str.c:689                                                                                                                                      
    #4 0x4183f6 in main ../test/unit/test_str.c:727                                                                                                                                           
    #5 0x7fca73c891e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)                                                                                                                         
    #6 0x40254d in _start (/home/rschiron/projects/rizinorg/rizin/build-strlcpy/test/unit/test_str+0x40254d)                                                                                  
                                                                                                                                                                                              
Address 0x7fffb3b5f99a is located in stack of thread T0 at offset 42 in frame                                                                                                                 
    #0 0x403674 in test_rz_str_ncpy ../test/unit/test_str.c:45

  This frame has 8 object(s):
    [32, 42) 'buf' (line 47)
    [64, 80) 'src' (line 58) <== Memory access at offset 42 partially underflows this variable
    [96, 5216) '_meqstr' (line 49) <== Memory access at offset 42 partially underflows this variable
    [5472, 10592) '_meqstr' (line 52)
    [10848, 15968) '_meqstr' (line 56)
    [16224, 21344) '_meqstr' (line 61)
    [21600, 26720) '_meqstr' (line 65)
    [26976, 32096) '_meqstr' (line 68)

```

**Comment by officialcjunior** [comment]
Yeah, looks like it fails in two places: when `size = sizeof(buf);` and when `n > src length` 💭

**Comment by XVilka** [comment]
If it will be fixed, it should be included in 0.3.0

File: meson.build [hunk 9]
```diff
@@ -369,7 +369,8 @@ foreach item : [
     ['system', '#include <stdlib.h>', []],
     ['fork', '#include <unistd.h>', []],
     ['nice', '#include <unistd.h>', []],
-    ['pipe2', '#define _GNU_SOURCE\n#include <fcntl.h>\n#include <unistd.h>', []]
+    ['pipe2', '#define _GNU_SOURCE\n#include <fcntl.h>\n#include <unistd.h>', []],
+    ['strlcpy', '#include <string.h>', []],
```
Comment (XVilka) [inline meson.build:374]: I recommend to swap the line with `pipe2` so the longest line will always be on the bottom.

File: librz/util/str.c [hunk 10]
```diff
@@ -753,22 +753,26 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
...
 
 	// do not do anything if n is 0
 	if (n == 0) {
 		return 0;
 	}
+#if HAVE_STRLCPY
+	return strlcpy(dst, src, n);
+#else
+	strncpy(dst, src, n);
+	dst[n] = '\0';
```
Comment (ret2libc) [inline librz/util/str.c]: This is probably where the issue is @officialcjunior . strlcpy (and rz_str_ncpy) takes the full size as argument, including the null byte. However you usually have to pass n-1 to strncpy and of course dst[n] goes out of bound because the last char is dst[n-1].

File: meson.build [hunk 11]
```diff
@@ -371,10 +371,11 @@ foreach item : [
     ['system', '#include <stdlib.h>', []],
     ['fork', '#include <unistd.h>', []],
     ['nice', '#include <unistd.h>', []],
+    ['strlcpy', '#include <string.h>', []],
     ['openpty', '', [utl]],
     ['forkpty', '', [utl]],
     ['login_tty', '', [utl]],
-    ['pipe2', '#define _GNU_SOURCE\n#include <fcntl.h>\n#include <unistd.h>', []]
+    ['pipe2', '#define _GNU_SOURCE\n#include <fcntl.h>\n#include <unistd.h>', []],
```
Comment (XVilka) [inline meson.build:378]: No need for comma here.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Comment by officialcjunior** [comment]
Over here, it looks like the test is fine while building with ASAN.

image

File: librz/util/str.c [hunk 12]
```diff
@@ -753,22 +753,25 @@ RZ_API char *rz_str_newf(const char *fmt, ...) {
...
-		dst[i] = src[i];
-	}
-	dst[i] = 0;
-	return i;
+#if HAVE_STRLCPY
+	return strlcpy(dst, src, n);
+#else
+	strncpy(dst, src, n - 1);
+	dst[n - 1] = '\0';
+	return n;
```
Comment (ret2libc) [inline librz/util/str.c]: This is not the correct return value. It should be the same as the one of strlcpy. Check the man page of strlcpy to understand what to return

**Review by ret2libc — CHANGES_REQUESTED** [review]

**Review by ret2libc — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]
